### ISSUE TITLE
Don't report CS8524 warning when extra cases were already handled outside the switch statement

### ISSUE THREAD

**Issue by IustinMicu** [dotnet/roslyn#49600]
**Brief description:**

The analyzer should not report the CS8524 warning in case the default switch case is not implemented in the switch but was already handled outside the switch, therefore the switch would be unreachable.

**Languages applicable:**

C#.
Maybe others, but I am not familiar with them.

**Code example that the analyzer should NOT report:**

```C#
public void Test(Iteration iteration)
{
    try
    {
        if (!Enum.IsDefined(typeof(Iteration), iteration))
        {
            return;
        }
    }
    catch
    {
        return;
    }

    // Currently CS8524 is reported here, because the default switch case was not handled (although it was handled above)
    // TODO: Don't report CS8524 in this situation
    var iterationPath = iteration switch
    {
        Iteration.Previous => "@CurrentIteration - 1",
        Iteration.Current => "@CurrentIteration"
    };
}

public enum Iteration
{
    Previous = -1,
    Current = 0
}
```

Here you can see the current situation where the warning is reported, although in any extra case the switch would not be reachable.

![analyzer](https://user-images.githubusercontent.com/22785067/100120671-e2c22d00-2e80-11eb-9cb9-01eee04cea8d.png)

**Comment by IustinMicu** [comment]
This is another example where this warning should not be reported:

```C#
public void Test(Iteration iteration)
{
    if (iteration != Iteration.Current && iteration != Iteration.Previous)
    {
        return;
    }

    // Currently CS8524 is reported here, because the default switch case was not handled (although it was handled above)
    // TODO: Don't report CS8524 in this situation
    var iterationPath = iteration switch
    {
        Iteration.Previous => "@CurrentIteration - 1",
        Iteration.Current => "@CurrentIteration"
    };
}

public enum Iteration
{
    Previous = -1,
    Current = 0
}
```

**Comment by RikkiGibson** [comment]
Generally the compiler's flow analysis only tracks things like "what values could be in this variable" in a very limited way. When a switch expression or switch statement is visited, there is no knowledge of what previous checks or assignments could have limited the possible set of values in the input to the switch, with the exception of `null` when nullable reference types is enabled.

You may be interested in the proposed language feature "closed enum types": dotnet/csharplang#3179
